test/executor_tests.cc

```diff
@@ -0,0 +1,939 @@
+
+	void outbound_pilot(const outbound_pilot& pilot) { m_executor->submit({}, {pilot}); }
+
+	void send(const node_id to_nid, const message_id msgid, const allocation_id source_aid, const range<3>& source_alloc_range, const id<3>& offset_in_alloc,
+	    const range<3>& send_range, const size_t elem_size) //
+	{
+		submit<send_instruction>(to_nid, msgid, source_aid, source_alloc_range, offset_in_alloc, send_range, elem_size);
+	}
+
+	/// Submits a host task that sleeps for the given duration.
+	void delay(const std::chrono::milliseconds& duration) {
+		submit<host_task_instruction>(
+		    [=](const box<3>& /* execution_range */, const communicator* /* collective_comm */) { std::this_thread::sleep_for(duration); },
+		    subrange(id<3>(0, 0, 0), range<3>(1, 1, 1)), range<3>(1, 1, 1), buffer_access_allocation_map{},
+		    collective_group_id(0) CELERITY_DETAIL_IF_ACCESSOR_BOUNDARY_CHECK(, task_type::host_compute, task_id(1), "task_name"));
+	}
+
+	/// Submits and awaits the shutdown epoch, then returns the collected operations log. Call after the last submission.
+	operations_log finish() {
+		CHECK(m_executor != nullptr);
+		epoch(epoch_action::shutdown, instruction_garbage{});
+		m_executor.reset();
+		return std::move(m_log);
+	}
+
+	task_id get_last_epoch() { return m_epochs.get(); }
+	task_id get_last_horizon() { return m_horizons.get(); }
+
+	void await_horizon(const task_id tid) { m_horizons.await(tid); }
+	void await_epoch(const task_id tid) { m_epochs.await(tid); }
+
+	const communicator* get_root_communicator() const { return m_root_comm; }
+
+  private:
+	instruction_id m_next_iid = 1;
+	std::optional<instruction_id> m_last_iid; // serialize all instructions for simplicity - we do not test OoO capabilities here
+	task_id m_next_task_id = task_manager::initial_epoch_task + 1;
+	std::vector<std::unique_ptr<instruction>> m_instructions; // we need to guarantee liveness as long as the executor thread is around
+	std::unique_ptr<executor> m_executor;
+	epoch_monitor m_horizons{0};
+	epoch_monitor m_epochs{0};
+	operations_log m_log;                      // mutated by executor thread - do not access before shutdown!
+	const communicator* m_root_comm = nullptr; // always nullptr for dry_run_executor
+
+	template <typename Instruction, typename... CtorParams>
+	instruction_id submit(CtorParams&&... ctor_args) {
+		const auto iid = instruction_id(m_next_iid++);
+		auto instr = std::make_unique<Instruction>(iid, 0, std::forward<CtorParams>(ctor_args)...);
+		if(m_last_iid.has_value()) { instr->add_dependency(*m_last_iid); }
+		m_executor->submit({instr.get()}, {});
+		m_instructions.push_back(std::move(instr));
+		m_last_iid = iid;
+		return iid;
+	}
+
+	void horizon_reached(const task_id tid) override { m_horizons.set(tid); }
+	void epoch_reached(const task_id tid) override { m_epochs.set(tid); }
+};
+
+
+TEST_CASE_METHOD(test_utils::dry_run_executor_fixture, "executors notify their delegate when encountering a horizon / epoch", "[executor]") {
```

Comment:
The only thing this fixture does is allow dry-run warnings, tests that only use `live_executor` don't have it. I agree this is not obvious. Maybe `dry_run_executor_warnings_fixture` or `any_executor_fixture`?

Reply:
I see - I think I'd prefer `dry_run_executor_warnings_fixture`. Although for this use case I think we could also just drop the `_fixture` suffix, maybe go for something like `suppress_dry_run_executor_warnings`.

---

Or... potential for overengineering?

```c++
	template <const char*... Warnings>
	class suppress_warnings {
	  public:
		suppress_warnings() { allow_higher_level_log_messages(detail::log_level::warn, std::string(Warnings)...); }
	};
```

Reply:
Yeah maybe this doesn't need to be a fixture at all since we don't care about the destructor. I've replaced `dry_run_executor_fixture` and `backend_fixture` with `test_utils::allow_*_warnings()` functions that are called at the beginning of each affected test and should communicate the intent better.